### PR TITLE
Johny fix bug create button earned dates for badges

### PR DESCRIPTION
**Pull request by johny4437**
# Description
Fixe bug ( PRIORITY MEDIUM) - 17
Need a “Details” button or other way to see when specific badges were earned. Some badges are earned multiple times, so the button needs to show all dates when they were earned

## Mainly changes explained:
1 -  added a  dropdown button  in the modal that show all earned badges
2 - Was added some lines of code in the badgeManagment.js file, to get the specified date when the badge is earned.

…

## How to test:
check into current branch
do `npm install` and `...` to run this PR locally
log as admin user (or any user)
go to dashboard→ scroll the screen→ Badges→ Badge Reports Button.
Will open a modal and will show the new functionality.
Captura de Tela (95)
[Video ](https://www.dropbox.com/s/8eylgss9bbe9iph/2022-12-31%2013-35-13.mp4?dl=0)

## Note:
To see this functionality you need to verify the pull request that I made for the backend too.

### CODE REVIEW

**Comment by one-community** [comment]
Excellent! Reminder to please also always put "done" and the PR # and link next to your name and the relevant bug on the bugs doc. See how I did this there for this one as an example.

**Comment by rauleffting** [comment]
I tested the new functionality - did a branch and npm install -, but, when I go to Badge Report, the page stays blank and the following errors appear in the console:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'length')
    at BadgeReport (BadgeReport.jsx:250:1)
    at renderWithHooks (react-dom.development.js:14803:1)
    at mountIndeterminateComponent (react-dom.development.js:17482:1)
    at beginWork (react-dom.development.js:18596:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188:1)
    at HTMLUnknownElement.sentryWrapped (helpers.ts:77:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237:1)
    at invokeGuardedCallback (react-dom.development.js:292:1)
    at beginWork$1 (react-dom.development.js:23203:1)
    at performUnitOfWork (react-dom.development.js:22154:1)
    at workLoopSync (react-dom.development.js:22130:1)
    at performSyncWorkOnRoot (react-dom.development.js:21756:1)
    at react-dom.development.js:11089:1
    at unstable_runWithPriority (scheduler.development.js:653:1)
    at runWithPriority$1 (react-dom.development.js:11039:1)
    at flushSyncCallbackQueueImpl (react-dom.development.js:11084:1)
    at flushSyncCallbackQueue (react-dom.development.js:11072:1)
    at discreteUpdates$1 (react-dom.development.js:21893:1)
    at discreteUpdates (react-dom.development.js:806:1)
    at dispatchDiscreteEvent (react-dom.development.js:4168:1)
    at HTMLDocument.sentryWrapped (helpers.ts:77:1)
```
```
The above error occurred in the <BadgeReport> component:
    in BadgeReport (created by Connect(BadgeReport))
    in Connect(BadgeReport) (at Badge.jsx:99)
    in div (created by ModalBody)
    in ModalBody (at Badge.jsx:98)
    in div (created by Modal)
    in div (created by Modal)
    in div (created by Transition)
    in Transition (created by Fade)
    in Fade (created by Modal)
    in div (created by Modal)
    in Portal (created by Modal)
    in Modal (at Badge.jsx:96)
    in div (created by CardBody)
    in CardBody (at Badge.jsx:80)
    in div (created by Card)
    in Card (at Badge.jsx:76)
    in div (created by Col)
    in Col (at Badge.jsx:75)
    in div (created by Row)
    in Row (at Badge.jsx:74)
    in div (created by Container)
    in Container (at Badge.jsx:73)
    in Badge (created by Connect(Badge))
    in Connect(Badge) (at Dashboard.jsx:84)
    in div (created by Col)
    in Col (at Dashboard.jsx:72)
    in div (created by Row)
    in Row (at Dashboard.jsx:68)
    in div (created by Container)
    in Container (at Dashboard.jsx:43)
    in Dashboard (created by Connect(Dashboard))
    in Connect(Dashboard) (at ProtectedRoute.jsx:29)
    in Route (at ProtectedRoute.jsx:21)
    in ProtectedRoute (created by Connect(ProtectedRoute))
    in Connect(ProtectedRoute) (at routes.js:143)
    in Switch (at routes.js:44)
    in Router (created by BrowserRouter)
    in BrowserRouter (at App.jsx:54)
    in PersistGate (at App.jsx:53)
    in Provider (at App.jsx:52)
    in App (at src/index.js:14)

React will try to recreate this component tree from scratch using the error boundary you provided, App.
```
```
index.js:1 Warning: App: Error boundaries should implement getDerivedStateFromError(). In that method, return a state update to display an error message or fallback UI.
    in App (at src/index.js:14)
```
```
Uncaught ReferenceError: process is not defined
    at 4043 (<anonymous>:2:13168)
    at r (<anonymous>:2:306599)
    at 8048 (<anonymous>:2:9496)
    at r (<anonymous>:2:306599)
    at 8641 (<anonymous>:2:1379)
    at r (<anonymous>:2:306599)
    at <anonymous>:2:315627
    at <anonymous>:2:324225
    at <anonymous>:2:324229
    at e.onload (index.js:1:1)
```
Screen Shot 2023-01-03 at 12 30 42

Please, feel free to correct me if did something wrong on testing it.

**Comment by rauleffting** [comment]
I tried to run your PR but it didn't work. I did branch and npm install in front and back. When I go to Badge Report the page stays white. I think the error is with the back-end. Here is what my console show in the back:

**Comment by rauleffting** [comment]
I tested your PR. The date feature to see when the same badges were earned worked well. However, I found a bug when trying to save changes. When clicking on 'save changes', the app crashes. Take a look at the video below:

https://user-images.githubusercontent.com/29555732/210402337-b2139d63-4483-453c-9b5b-78322cb323e0.mov

**Comment by rauleffting** [comment]
Thank you very much, Johny. First, I'd like to apologize for not reporting the problem properly. My fault.
I told you about the bug because it is inside your PR. The current version of the app, the one we are using, doesn't crash when clicking on save changes. If your PR is approved, it will solve the date problem but also creates a new bug. However, if Mr. Jae agrees, we can approve it and add the 'save changes' bug to the bug list, create a new PR, etc... Guess you and him will know what's the better option. Please, feel free to correct me if I'm wrong.

**Comment by rauleffting** [comment]
Thank you very much, Johny. First, I'd like to apologize for not reporting the problem properly. My fault.
I told you about the bug because it is inside your PR. The current version of the app, the one we are using, doesn't crash when clicking on save changes. If your PR is approved, it will solve the date problem but also creates a new bug. However, if Mr. Jae agrees, we can approve it and add the 'save changes' bug to the bug list, create a new PR, etc... Guess you and him will know what's the better option. Please, feel free to correct me if I'm wrong.

@one-community

**Comment by rauleffting** [comment]
Please, take a look again. The errors are not all the same. 

**(Development Branch)**
Below is the development branch:

https://user-images.githubusercontent.com/29555732/210801079-66a2a3bc-4658-4d0d-b5b1-ad369d7a0658.mov

As you can see, the page doesn't freeze. And it's only one error that appears in the console. You can close the popup and still navigate. Here is a picture of the development branch error:

error1

**(PR Branch)**
Now, please, look at the error in the PR:

https://user-images.githubusercontent.com/29555732/210802940-3e70524e-59af-4586-a0bc-ed3a6e8ed69a.mov

As you can see, the page freezes. And it's also needed to reload the page to get the app working again. Here is a picture of the PR errors:

error2

**Comment by johny4437** [comment]
Hi @rauleffting  thank you. So what I did? I forked the development branch again(backend and frontend), because I wanted to verify if this error was just in this PR or was in the development branch too, so I tested here and this error is already in the development branch(updated), I took a look at the code, and the function `props.setUserProfile( )`  does not exist,  so I made a video testing in the development branch(updated) and  the app is freezing when are made some changes in the Badge Report, I am sorry because I did not pay atention when I was resolving this PR.
Captura de Tela (132)

https://user-images.githubusercontent.com/39320057/211144429-fc295d7b-4bc7-4705-8dc7-4603912c9baf.mp4

**Comment by rauleffting** [comment]
You are right Johny! Thank you very much. I wasn't testing the development branch locally like you, but with the hgnapplication_react_dev.surge.sh. It's strange because in this one the application doesn't freeze. I don't know why this is happening. That's why we weren't understanding each other. However, I'm going to approve it.

**Review by rauleffting — DISMISSED** [review]
It has been revised several times and Johny has made this part of the code cleaner.
Screen Shot 2023-01-07 at 16 46 29

**Comment by EvianTan** [comment]
Hey @johny4437 @rauleffting ,
I was trying to test this PR, I run both your frontend/backend PR but got nothing when I try to assign badges to myself. I post my video here, please tell me if any step I missed - [video](https://www.dropbox.com/s/awxvwgrkqgdawzm/test%20Jonny%20badge%20PR%20but%20don%27t%20work.mp4?dl=0)

File: src/actions/badgeManagement.js [hunk 1]
```diff
@@ -113,18 +113,46 @@ export const assignBadges = (firstName, lastName, selectedBadges) => {
     }
     const badgeCollection = res.data[0].badgeCollection;
     const UserToBeAssigned = res.data[0]._id;
+    const earnedDate = badgeCollection[0].earnedDate;
```
Comment (EvianTan) [inline src/actions/badgeManagement.js:116]: No functional changes above this line.

File: src/components/Badge/Badge.jsx [hunk 2]
```diff
@@ -67,50 +67,51 @@ const Badge = props => {
   }, [props.userProfile.badgeCollection, totalBadge]);
 
   const permissionsUser = props.userProfile?.permissions?.frontPermissions;
+  console.log('ALL USER BADGE ==>', props.userProfile.badgeCollection);
```
Comment (EvianTan) [inline src/components/Badge/Badge.jsx:70]: Please remove this debug output line

File: src/components/UserProfile/UserProfile.jsx [hunk 3]
```diff
@@ -40,7 +40,6 @@ import hasPermission from 'utils/permissions';
 import ActiveInactiveConfirmationPopup from '../UserManagement/ActiveInactiveConfirmationPopup';
 import { updateUserStatus } from '../../actions/userManagement';
 import { UserStatus } from '../../utils/enums';
-import { faSleigh } from '@fortawesome/free-solid-svg-icons';
```
Comment (EvianTan) [inline src/components/UserProfile/UserProfile.jsx:43]: Nice catch!

File: src/components/SummaryBar/SummaryBar.jsx [hunk 4]
```diff
@@ -84,7 +84,10 @@ const SummaryBar = props => {
   useEffect(() => {
     loadUserProfile();
     getUserTask();
-  }, [leaderData]);
```
Comment (EvianTan) [inline src/components/SummaryBar/SummaryBar.jsx:87]: Could you help me understand why `leaderData` removed here?

File: src/components/Projects/WBS/WBSDetail/AddTask/AddTaskModal.jsx [hunk 5]
```diff
@@ -307,8 +305,22 @@ const AddTaskModal = props => {
     setEndstateInfo(props.tasks.copiedTask.endstateInfo);
   };
 
+  //FUNCTION TO UPDATE TASK MOTHER
+  const updateTaskMother = () => {
```
Comment (EvianTan) [inline src/components/Projects/WBS/WBSDetail/AddTask/AddTaskModal.jsx:309]: Changes in this file are not related to this PR, please revert them. We should always follow "One PR only do (a part of) one thing" rule.

File: src/components/Badge/BadgeReport.jsx [hunk 6]
```diff
@@ -245,6 +247,12 @@ const BadgeReport = props => {
     //Reload the view profile page with updated bages
     window.location.reload();
   };
+  console.log(props.badges[1].earnedDate.length);
+  let arr = [];
```
Comment (EvianTan) [inline src/components/Badge/BadgeReport.jsx:251]: Please remove your new line 250-255. Looks like the `arr` only used in this snippet of code. Please correct me if I'm wrong. 
image

File: src/components/Badge/Badge.jsx [hunk 7]
```diff
@@ -67,50 +67,51 @@ const Badge = props => {
...
   return (
     <>
-    <Container>
```
Comment (EvianTan) [inline src/components/Badge/Badge.jsx:72]: No functional changes under this line.

File: src/actions/badgeManagement.js [hunk 8]
```diff
@@ -113,18 +113,46 @@ export const assignBadges = (firstName, lastName, selectedBadges) => {
...
           badgeObj.lastModified = Date.now();
+          badgeObj.earnedDate = [...earnedDate, formatedDate];
           included = true;
         }
       });
       if (!included) {
-        badgeCollection.push({ badge: badgeId, count: 1, lastModified: Date.now() });
+        const today = new Date();
+        const yyyy = today.getFullYear();
+        let mm = today.getMonth();
```
Comment (EvianTan) [inline src/actions/badgeManagement.js:140]: Missed `+1` at the end of this line.

File: src/actions/badgeManagement.js (hunk 1 above)
Comment (johny4437) [reply src/actions/badgeManagement.js:116]: were not added by me

File: src/actions/badgeManagement.js (hunk 8 above)
Comment (johny4437) [reply src/actions/badgeManagement.js]: Done!

File: src/components/Badge/BadgeReport.jsx (hunk 6 above)
Comment (johny4437) [reply src/components/Badge/BadgeReport.jsx]: Yes you are right, removed!

**Comment by Filipe-Oliv** [comment]
Hi @johny4437 

I don't know if this bug is related to your PR, but have a look at it. If not let kow that I'll report them.

When I changed the page resolution the 'table' was flexible but its child items were not.

[week 03 - 02-09-23 (#265).webm](https://user-images.githubusercontent.com/114119668/217857297-3977ba7a-6567-447e-a5b2-0e5b32a39108.webm)

**Review by kamtoeddy — COMMENTED** [review]
Everything works and the code looks ok but I noticed the same problem @Filipe-Oliv points out [here](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/631#issuecomment-1424413517)

I have checked on the development branch and it seems the problem is from this PR

**Comment by PrabhjotSembhi** [comment]
@johny4437 Whats the status of this? is all comments are resolved

**Comment by johny4437** [comment]
@Filipe-Oliv and @kamtoeddy I will fix this thank you for the review!

**Comment by johny4437** [comment]
@Filipe-Oliv and @kamtoeddy now the table is responsive.

**Comment by johny4437** [comment]
Guys I noticed something here, the earned dates are being saved when a badge is assigned manually, but  when a badge is assigned automaticaly the earned dates are not being  saved, I will fix this in the backend PR.

**Comment by Filipe-Oliv** [comment]
Hi @johny4437 

Much better, but just one thing. The box for 'Count' is not fitting well at a certain resolution. Take a look at the video below.

[02-22-2023 (1).webm](https://user-images.githubusercontent.com/114119668/220600709-77924a68-b16e-46b2-85a7-1699ad4e1ef0.webm)

Waiting for the 'badge will be assigned automatically dates will not be saved' solution.

**Comment by Lunaliar** [comment]
Hey @johny4437, I tried to review this but every time I click on the badge report button the app crashes and goes white. I was able to figure out that it's having issues reading data from the sortBadges state but I'm having problems debugging it. Has anyone else run into this?
[Screencast from 02-23-2023 11:14:01 AM.webm](https://user-images.githubusercontent.com/98275089/220966114-08758e9e-577c-485e-a235-253276921071.webm)

**Review by kamtoeddy — APPROVED** [review]

File: src/components/Badge/BadgeReport.jsx [hunk 9]
```diff
@@ -328,6 +331,19 @@ const BadgeReport = props => {
...
                   []
                 )}
+                <td>
+                  <Dropdown>
+                    <Dropdown.Toggle variant="secondary" id="dropdown-basic">
+                      Earned Dates
+                    </Dropdown.Toggle>
+
+                    <Dropdown.Menu>
+                      {value.earnedDate.map((item, i) => (
```
Comment (Lunaliar) [inline src/components/Badge/BadgeReport.jsx:341]: Hey, @johny4437 while looking trying to review this the badge report modal window would crash the app. I think I tracked it down to this line, value does not have an earnedDate property. I removed "value." from the line and was able to see the modal window, however interacting with this dropdown menu would also crash the app.

**Review by Lunaliar — COMMENTED** [review]
This commit kept crashing the app for me, after some debugging I think I tracked it down to line 341 in BadgeReport, can you please take a look and make sure it's still working for you?

**Comment by ayush0904** [comment]
Hi @johny4437, as mentioned above the app is crashing when you go to the other person's dashboard and click the Badge Report button. However, the Badge Report button is working perfectly when a person is in his own dashboard.

**Review by tingedudu8 — CHANGES_REQUESTED** [review]
Hi @johny4437, there are two actions will lead to the whole app crush, the first one is click on the "report" button in other people's dashboard, and another is to save the change. You can see the error message in the video. Also, the bug needs to fix is that the badges assigned automatically will not leave a record for their assigned time. Thanks!

https://user-images.githubusercontent.com/70128933/221749466-649bd756-3931-46c4-bf9d-84563307bca1.mov

**Comment by NyahJ** [comment]
Hi @johny4437,

I tried to test the functional feature change made by this PR, seeing if there is any update.
As mentioned above, the crash still exists, could you pls take a look into that? Thanks
截屏2023-03-01 下午4 31 22

File: src/components/Badge/BadgeReport.jsx (hunk 9 above)
Comment (johny4437) [reply src/components/Badge/BadgeReport.jsx:341]: You need to test this branch with the backend branch too, if youy don't,  won't work.

**Comment by mpavithr** [comment]
Hi, I tried testing this, and i have pr265 running as the backend, and I do see the new column of earned dates
pr631_1
However, when I try to assign badges to try and test whether they will be reflected in the report, I am unable to assign any badges, not sure if this is under the scope of this PR
pr265_1
